cpp/src/arrow/ipc/writer.cc

```diff
@@ -1147,13 +1201,16 @@ Status SerializeSchema(const Schema& schema, MemoryPool* pool,
   std::shared_ptr<io::BufferOutputStream> stream;
   RETURN_NOT_OK(io::BufferOutputStream::Create(1024, pool, &stream));
 
-  DictionaryMemo memo;
-  SchemaWriter schema_writer(schema, &memo, pool, stream.get());
+  // XXX: PayloadStreamWriter takes a std::shared_ptr<Schema> but we only got a
+  // raw reference.  We fake it with a null deleter.
+  auto deleter = [](Schema*) -> void {};
+  std::shared_ptr<Schema> schema_ptr(const_cast<Schema*>(&schema), deleter);
```

Comment:
This may work but it's highly irregular. For clarity it might be worthwhile to make a deep copy of the schema:
```suggestion
  auto schema_ptr = arrow::schema(schema.fields(), schema.metadata());
```